src/main/java/org/zeroturnaround/zip/ZipEntryUtil.java

```diff
@@ -78,20 +81,90 @@ static void copyEntry(ZipEntry zipEntry, InputStream in, ZipOutputStream out) th
    * Copies a given ZIP entry to a ZIP file. If this.preserveTimestamps is true, original timestamp
    * is carried over, otherwise uses current time.
    *
-   * @param zipEntry
+   * @param originalEntry
    *          a ZIP entry from existing ZIP file.
    * @param in
    *          contents of the ZIP entry.
    * @param out
    *          target ZIP stream.
    */
-  static void copyEntry(ZipEntry zipEntry, InputStream in, ZipOutputStream out, boolean preserveTimestamps) throws IOException {
-    ZipEntry copy = copy(zipEntry);
-    copy.setTime(preserveTimestamps ? zipEntry.getTime() : System.currentTimeMillis());
+  static void copyEntry(ZipEntry originalEntry, InputStream in, ZipOutputStream out, boolean preserveTimestamps) throws IOException {
+    ZipEntry copy = copy(originalEntry);
+
+    if (preserveTimestamps) {
+      /**
+       * If we succeed to set the modified time (works only on JDK8)
+       * then we shouldn't call the setTime method anymore as that
+       * will nullify the modified time - see
+       * http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/f1f3f9eaf7fa/src/share/classes/java/util/zip/ZipEntry.java#l163
+       */
+      boolean success = tryToPreserveLatestJDKSupportedTimes(copy, originalEntry);
+      if (!success) {
+        copy.setTime(originalEntry.getTime());
+      }
+    }
+    else {
+      copy.setTime(System.currentTimeMillis());
+    }
     addEntry(copy, new BufferedInputStream(in), out);
   }
 
   /**
+   * The method tries to set the last modified time. The pre-requisite for this is that
```

Comment:
Some problems with this code:
* it uses reflection over-and-over again, in a potentially performance-sensitive place (this code gets executed for each `ZipEntry`)
* it does not belong in this class

Better approach would be to hide the timestamp setting behind a (strategy) interface, provide two implementations of it and then, figure out once (at class loading time) which implementation to use and store that implementation in a static field.

For example, see `ZTFilePermissionsStrategy` interface and it's usage.